Sources/Basics/HTTPClient/URLSessionHTTPClient.swift

```diff
@@ -202,18 +254,61 @@ private class DataTaskManager: NSObject, URLSessionDataDelegate {
         init(
             task: URLSessionDataTask,
             progressHandler: LegacyHTTPClient.ProgressHandler?,
+            dataTaskManager: DataTaskManager,
             completionHandler: @escaping LegacyHTTPClient.CompletionHandler,
             authorizationProvider: LegacyHTTPClientConfiguration.AuthorizationProvider?
         ) {
             self.task = task
             self.progressHandler = progressHandler
+            self.dataTaskManager = dataTaskManager
             self.completionHandler = completionHandler
             self.authorizationProvider = authorizationProvider
         }
     }
 }
 
-private class DownloadTaskManager: NSObject, URLSessionDownloadDelegate {
+/// This uses the same pattern as `WeakDataTaskManager`. See comment on that type.
+private class WeakDownloadTaskManager: NSObject, URLSessionDownloadDelegate {
+    private weak var downloadTaskManager: DownloadTaskManager?
+
+    init(_ downloadTaskManager: DownloadTaskManager? = nil) {
+      self.downloadTaskManager = downloadTaskManager
+    }
+
+    func urlSession(
+        _ session: URLSession,
+        downloadTask: URLSessionDownloadTask,
+        didWriteData bytesWritten: Int64,
+        totalBytesWritten: Int64,
+        totalBytesExpectedToWrite: Int64
+    ) {
+        downloadTaskManager?.urlSession(
+            session,
+            downloadTask: downloadTask,
+            didWriteData: bytesWritten,
+            totalBytesWritten: totalBytesWritten,
+            totalBytesExpectedToWrite: totalBytesExpectedToWrite
+        )
+    }
+
+    func urlSession(
+        _ session: URLSession,
+        downloadTask: URLSessionDownloadTask,
+        didFinishDownloadingTo location: URL
+    ) {
+        downloadTaskManager?.urlSession(session, downloadTask: downloadTask, didFinishDownloadingTo: location)
+    }
+
+    public func urlSession(
```

Comment:
`public` function in a `private` type

Reply:
FWIW I see nothing wrong with `public` visibility inside a `private` type, ie. visibility inside the type is describing the intent of the API and then visibility on the type for how accessible you want it to be. It makes moving (and re-using) that type later much simpler. One argument against that is it's quite likely a private type hasn't had much thought to its API, which could be true 🤷.